src/control/server/mgmt_client.go

```diff
@@ -59,6 +59,26 @@ func newMgmtSvcClient(ctx context.Context, log logging.Logger, cfg mgmtSvcClient
 	}
 }
 
+// delayRetry delays next retry.
+func (msc *mgmtSvcClient) delayRetry(ctx context.Context) {
+	delayCtx, cancelDelayCtx := context.WithTimeout(ctx, retryDelay)
+	select {
+	case <-delayCtx.Done():
+	}
+	cancelDelayCtx()
+}
+
+func (msc *mgmtSvcClient) logAttempt(prefix string, err error, status *int32) {
```

Comment:
This feels like it should be split into separate functions, since the behavior is totally different and uses different parameters depending on which case it is. I think you could eliminate the sort of confusing int32 pointer passing that way, too.

Reply:
I was initially asked to remove duplication from the Join and Stop methods hence the current implementation but I will revert